rbac/management/group/definer.py

```diff
@@ -92,10 +93,14 @@ def set_system_flag_before_update(group: Group, tenant, user) -> Optional[Group]
 
 def clone_default_group_in_public_schema(group, tenant) -> Optional[Group]:
     """Clone the default group for a tenant into the public schema."""
-    if settings.PRINCIPAL_CLEANUP_UPDATE_ENABLED_UMB:
-        # TODO: bootstrap the tenant to get the mapping
-        # use this for uuid instead and to remove the default role binding tuple
-        group_uuid = uuid4()
+    if settings.PRINCIPAL_CLEANUP_UPDATE_ENABLED_UMB and settings.REPLICATION_TO_RELATION_ENABLED:
+        tenant_bootstrap_service = V2TenantBootstrapService(OutboxReplicator())
+        bootstrapped_tenant = tenant_bootstrap_service.bootstrap_tenant(tenant)
```

Comment:
If after my CJI job run, I think yes. But now it is not guaranteed a tenant has those workspaces created.
However, it is ok either the tenant has or not. It is guaranteed to coexist with the tenant mapping. If the tenant mapping exist, so does the default/root ws. If not, we create them

Reply:
Right, by definition a tenant is not considered "bootstrapped" (in v2) if they don't have built in workspaces. That's part of the process. So `bootstrap_tenant` must create the builtin workspaces as well as mapping.

>         # This assumes that Tenant Mapping exist

Not sure what is meant by the comment exactly – but my interpretation of it is that it is maybe misleading. The method below _does not_ assume the TenantMapping exists. It creates it if it doesn't.

Reply:
This method is called only from API endpoint and in that situation TenantMapping has to be created already. 
But maybe in some edge cases it is not true ? 

I can remove the comment. My main concern was whether the tenant can be created, but the default workspace is not, or if the tenant mapping could be created, but the default workspace is not. Because if the tenant mapping is not created and the default workspace is already created, this method will fail - but it looks is not possible.

Reply:
> This method is called only from API endpoint and in that situation TenantMapping has to be created already. But maybe in some edge cases it is not true ?

Okay, yeah I see the issue. We need to update the settings check above to check for V2_BOOTSTRAP_TENANT I think. I think this would only happen during migration steps as of latest doc. There is a window where a default group may attempt to be created for an existing tenant that is not yet bootstrapped.

We could keep this as is, and then I think it should never happen to your point. The user import job will take care of race conditions in that case. But, I think it is a little safer to bootstrap the tenant in that case, to ensure we use the uuid from the tenant mapping. And we know we cannot get into trouble there, because of the 1-1 unique constraint on the tenant mapping, any concurrent tx creating one (e.g. the user import) will fail.

> My main concern was whether the tenant can be created, but the default workspace is not, or if the tenant mapping could be created, but the default workspace is not. Because if the tenant mapping is not created and the default workspace is already created, this method will fail - but it looks that it is not possible.

Yes, correct – we should never have workspaces without the TenantMapping. That's why the bootstrap service stuff was created – to consolidate the logic to enforce those kinds of things.